src/content/blog/why-typed-linting-needs-typescript-today/index.mdx

```diff
@@ -0,0 +1,135 @@
+
+If type checking is so important, and a native speed type checker would be so beneficial, why hasn't one been written yet?
+Why hasn't anybody reimplemented TypeScript in, say, Go or Rust?
+
+TypeScript is a huge project under active development from a funded team of incredibly dedicated, experienced Microsoft employees -- as well as an active community of power users and contributors.
+The TypeScript team receives the equivalent of _millions of dollars a year_ in funding from employee compensation alone.
+A new version of TypeScript that adds type checking bugfixes and features releases every three months.
+
+Can you imagine the Herculean difficulty for any team trying to keep up with TypeScript?
+
+I hope for a day when there is a tool that can fully compete with TypeScript.
+Competition is good for an ecosystem.
+But it's going to be _years_ until a tool like that can develop.
+
+#### Subset Reimplementation
+
+TypeScript consists of several big areas of functionality, of which type checking is only one.
+"Type checking" as used by tooling primarily consists of two areas of APIs:
+
+- Type retrieval _("what type is this node?")_
+- Type relations _("is this type assignable to that type?")_
+
+Most typed lint rules today only use the _retrieval_ part of the type checker.
+Reimplementing only that portion of TypeScript could significantly reduce the development cost of the reimplementation.
+
+Unfortunately, the _relations_ portion of TypeScript was recently confirmed to be ready for use by linters[^tseslint-discussion-checker-is-type-assignable-to].
+typescript-eslint is going to start building in rule logic that builds on `checker.isTypeAssignableTo`.
+
+Any subset reimplementation of TypeScript would need to include equivalents both for the existing type _retrieval_ and type _relations_ APIs.
+Those APIs include most of the tricky assignability logic that makes TypeScript so hard to reimplement in the first place.
+
+A subset reimplementation of TypeScript for typed linting is tempting, but not as narrowly scoped of a task as it might seem.
+
+### Automatically Porting TypeScript to Native Speed
+
+> I previously touched on similar points in [Rust-Based JavaScript Linters: Fast, But No Typed Linting Right Now > Option: Boosting TypeScript’s APIs to Native Speed](/blog/rust-based-javascript-linters-fast-but-no-typed-linting-right-now/#option-boosting-typescripts-apis-to-native-speed).
+
+Instead of reimplementing TypeScript from scratch, what if we could automatically port its source code to a faster language?
+
+Well, TypeScript has tens of thousands of lines of source code built for single-threaded JavaScript with dynamic objects.
+It's really not architected to work in another paradigm.
+I've yet to see anybody make any significant progress on this idea.
```

Comment:
Actually, TypeRunner could be considered "significant" progress. I'll have to rephrase.